python/mlcroissant/mlcroissant/_src/structure_graph/nodes/file_object.py

```diff
@@ -31,30 +34,44 @@ class FileObject(Node):
     sha256: str | None = None
     source: Source | None = None
 
-    def __post_init__(self):
-        """Checks arguments of the node."""
+    def __post_init__(self, uuid: str | None = None):
+        """Checks arguments of the node and sets UUID."""
+        if not uuid:
+            uuid = generate_uuid()
+        self._uuid = uuid
         self.validate_name()
-        self.assert_has_mandatory_properties("encoding_format", "name")
+        self.assert_has_mandatory_properties("encoding_format", "name", "_uuid")
         if not self.contained_in:
             self.assert_has_mandatory_properties("content_url")
             if self.ctx and not self.ctx.is_live_dataset:
                 self.assert_has_exclusive_properties(["md5", "sha256"])
 
     def to_json(self) -> Json:
         """Converts the `FileObject` to JSON."""
-        return remove_empty_values({
+        contained_in = unbox_singleton_list(self.contained_in)
+        if not self.ctx.is_v0():
+            if isinstance(contained_in, list):
+                contained_in = [{"@id": uuid_to_jsonld(source)} for source in contained_in]  # type: ignore
```

Comment:
Do this before unbox_singleton_list to avoid the type error

Reply:
Right, done.